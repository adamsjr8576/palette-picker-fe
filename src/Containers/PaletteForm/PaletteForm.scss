.palette-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;
  border: 3px solid black;
  padding: 15px;
  width: 80%;
  border-radius: 20px;
  height: 200px;
  background-color: #F2F2F2;
}

.palette-form-header {
  font-size: 1.5em;
  margin: 0;
}

.palette-form-label {
  font-size: 1.5em;
}

.project-selector-container {
  display: flex;
  align-items: center;
  justify-content: space-around;
  width: 90%;
}

.project-selector {
  font-size: 1em;
  padding: 10px;
}

.palette-name-input {
  width: 70%;
  font-size: 1em;
  padding: 10px;
  border-radius: 10px;
  outline: none;
  border: 2px solid black;
}

.select-palette-btn {
  font-size: 1.25em;
  padding: 10px;
  border-radius: 10px;
  cursor: pointer;
  border: 0;
  background-color: #444343;
  color: white;
  outline: none;
}

.select-palette-btn:hover {
  box-shadow: 3px 5px #203246;
}

.select-palette-btn:focus {
  transform: scale(1.05);
}

.select-palette-btn:active {
  transform: translate(3px, 3px);
  box-shadow: 1px 2px #203246;
}
