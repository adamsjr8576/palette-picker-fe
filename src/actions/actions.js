export const addProject = (project) => ({
  type: 'ADD_PROJECT',
  project
})