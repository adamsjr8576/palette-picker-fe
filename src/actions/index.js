export const addNewPalette = currentPalette => ({
  type: 'ADD_PALETTE',
  currentPalette
});

export const updatePaletteLocked = hexCode => ({
  type: 'UPDATE_PALETTE_LOCKED',
  hexCode
});

export const addInitialProjects = projects => ({
  type: 'ADD_INITIAL_PROJECTS',
  projects
});

export const addPalettes = palettes => ({
  type: 'ADD_PALETTE_TO_PROJECT',
  palettes
});

export const addProject = project => ({
  type: 'ADD_PROJECT',
  project
});

export const deletePaletteFromStore = palette => ({
  type: 'DELETE_PALETTE',
  palette
});

export const deleteProjectFromStore = project => ({
  type: 'DELETE_PROJECT',
  project
});
