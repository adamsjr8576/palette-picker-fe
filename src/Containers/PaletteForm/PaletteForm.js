import React, { useState } from 'react';
import { connect } from 'react-redux';
import { addPalettes } from '../../actions/index';
import './PaletteForm.scss';
import { postPalette, getPaletteById } from '../../apiCalls';

export const PaletteForm = ({ addPalettes, projects, currentPalette }) => {
  const [ newProject, selectNewProject ] = useState('');
  const [ paletteName, setPalette ] = useState('');
  const [ error, setError ] = useState(false);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (newProject === '' || paletteName === '') {
      setError(true);
      resetInputs();
    } else {
      postNewPalette(paletteName, currentPalette);
      resetInputs();
      setError(false);
    }
  }

  const resetInputs = () => {
    selectNewProject('');
    setPalette('');
  }

  const createPostBody = () => {
    const selectedProject = projects.find(project => project.name === newProject);
    const projectId = selectedProject.id;
    return currentPalette.reduce((acc, color, index) => {
      if (index === 0) {
        acc.color_one = color.color;
      }
      if (index === 1) {
        acc.color_two = color.color;
      }
      if (index === 2) {
        acc.color_three = color.color;
      }
      if (index === 3) {
        acc.color_four = color.color;
      }
      if (index === 4) {
        acc.color_five = color.color;
      }
      return acc;
    }, {name: paletteName, project_id: projectId });
  }

  const postNewPalette = (paletteName, currentPalette) => {
    const body = createPostBody()
    postPalette(body)
      .then(id => {
        getPaletteById(id)
          .then(palette => {
            addPalettes(palette);
          })
          .catch(err => console.log(error));
      })
      .catch(err => console.log(error));
  }

  const createProjectOptions = (newProject) => {
    return projects.map(project => {
      return (
        <option selected={newProject === project.name} value={project.name} key={project.name}>
          {project.name}
        </option>
      );
    });
  }

  let outlineColor = '#017DF1';
  if (currentPalette.length) {
    outlineColor = currentPalette[0].color;
  }

  const options = createProjectOptions(newProject);
  return(
    <form className='palette-form'>
      <p className='palette-form-header'>Add Palette To Project: </p>
      <div className="project-selector-container">
        <label for='select-project' className='palette-form-label'>Select Project: </label>
        <select
          id='select-project'
          data-testid='select-project'
          className='project-selector'
          value={newProject}
          onChange={ e => selectNewProject(e.target.value) }
        >
          <option value="">Please Select</option>
          {options}
        </select>
      </div>
      <input
        data-testid='palette-name-input'
        placeholder='Palette Name'
        className='palette-name-input'
        type='text'
        maxLength='30'
        value={paletteName}
        onChange={ e => setPalette(e.target.value) }
        style={{border: `3px solid ${outlineColor}`}}
      />
      <div className='project-form-error-container'>
        <p data-testid='project-form-error' className='project-form-error' hidden={!error}>Error: Please Enter a Name and Project</p>
      </div>
      <button className='select-palette-btn' type='button' onClick={ e => handleSubmit(e) }>Save Palette</button>
    </form>
  )
}

export const mapDispatchToProps = dispatch => ({
  addPalettes: palette => dispatch( addPalettes(palette) )
});

export const mapStateToProps = state => ({
  projects: state.projects,
  currentPalette: state.currentPalette
});

export default connect(mapStateToProps, mapDispatchToProps)(PaletteForm);
