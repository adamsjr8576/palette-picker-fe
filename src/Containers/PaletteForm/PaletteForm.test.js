import React from 'react';
import { shallow } from 'enzyme';
import { PaletteForm, mapStateToProps, mapDispatchToProps } from './PaletteForm';
import { addPalettes } from '../../actions/index';
import { render, fireEvent } from '@testing-library/react';

describe('PaletteForm', () => {

  let mockProps, setup;

  beforeEach(() => {

    mockProps = {
      addPalettes: jest.fn(),
      projects: [{id: 1, name: "test hahahah", created_at: "2020-02-07T20:06:59.135Z", updated_at: "2020-02-07T20:06:59.135Z"},
        {id: 2, name: "test", created_at: "2020-02-07T20:06:59.135Z", updated_at: "2020-02-07T20:06:59.135Z"}],
      currentPalette: [
        {locked: false, color: '#11111'},
        {locked: false, color: '#11111'},
        {locked: false, color: '#11111'},
        {locked: false, color: '#11111'},
        {locked: false, color: '#11111'}]
    }

    setup = () => {
      const wrapper = render(<PaletteForm
        addPalettes={mockProps.addPalettes}
        projects={mockProps.projects}
        currentPalette={mockProps.currentPalette}
      />);
      const select = wrapper.getByTestId('select-project')
      const input = wrapper.getByTestId('palette-name-input')
      const btn = wrapper.getByRole('button')
      const error = wrapper.getByTestId('project-form-error')
      return {
        select,
        input,
        btn,
        error,
        wrapper
      }
    }
  });

  describe('PaletteForm Component', () => {

    test('should match the snapshot', () => {
      const { wrapper } = setup();

      expect(wrapper).toMatchSnapshot();
    });

    test("should load with initial state of '' in the add palette input", () => {
      const { input } = setup();

      expect(input.textContent).toBe('');
    });

    test('should allow a user to input a palette name', () => {
      const { input } = setup();

      fireEvent.change(input, { target: { value: 'New Palette Name'} });

      expect(input.value).toBe('New Palette Name');
    });

    test('should reset the input when button is clicked', () => {
      const { select, input, btn } = setup();

      const inputValue = input;

      fireEvent.change(select, {target: {value: 'test'}})
      fireEvent.change(inputValue, { target: { value: 'New Project Name'} });
      fireEvent.click(btn);

      expect(inputValue.textContent).toBe('');
    });

    test('should select a project from the dropdown', () => {
      const { select, wrapper } = setup();

      fireEvent.change(select, {target: {value: 'test'}});

      expect(wrapper.getByTestId('select-project').value).toBe('test');
      expect(wrapper.getByDisplayValue('test')).toBeInTheDocument();
    });

    test('should run setError and display errorif button is pushed with nothing in inputs', () => {
      const { btn, error, wrapper, input } = setup();

      fireEvent.click(btn);

      expect(error.hidden).toEqual(false);
      expect(input.textContent).toBe('');
    });

    test('should run setError and display errorif button is pushed with no project selected', () => {
      const { btn, error, wrapper, select, input } = setup();

      fireEvent.change(select, {target: {value: 'test'}});
      fireEvent.click(btn);

      expect(error.hidden).toEqual(false);
      expect(input.textContent).toBe('');
    });

    test('should run setError and display errorif button is pushed with no palette name', () => {
      const { btn, error, wrapper, select, input } = setup();

      fireEvent.change(input, { target: { value: 'New Project Name'} });
      fireEvent.click(btn);

      expect(error.hidden).toEqual(false);
      expect(input.textContent).toBe('');
    });

  });

  describe('mapDispatchToProps', () => {

    test('should call mapDispatch with a palette when handleSubmit is called', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = addPalettes({
        name: "Palette Blue",
        project_id: 2,
        color_one: "#87085",
        color_two: "#87085",
        color_three: "#87085",
        color_four: "#87085",
        color_five: "#87085"
      });

      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.addPalettes({
        name: "Palette Blue",
        project_id: 2,
        color_one: "#87085",
        color_two: "#87085",
        color_three: "#87085",
        color_four: "#87085",
        color_five: "#87085"
      });

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });

  describe('mapStateToProps', () => {
    it('should return an object with the currentPalette and projects data', () => {
      const mockState = {
        currentPalette: [
          {locked: false, color: '#444343'},
          {locked: false, color: '#6D6D6D'},
          {locked: false, color: '#9B9B9B'},
          {locked: false, color: '#C2C2C2'},
          {locked: false, color: '#DCDCDC'}],
        projects: [{name: 'test project', id: 4}],
        palettes: []
      };
      const expected = {
        currentPalette: [
          {locked: false, color: '#444343'},
          {locked: false, color: '#6D6D6D'},
          {locked: false, color: '#9B9B9B'},
          {locked: false, color: '#C2C2C2'},
          {locked: false, color: '#DCDCDC'}],
          projects: [{name: 'test project', id: 4}],
      };

      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

});
