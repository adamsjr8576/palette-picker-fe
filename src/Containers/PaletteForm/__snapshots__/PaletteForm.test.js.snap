// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteForm PaletteForm Component should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="palette-form"
      >
        <p
          class="palette-form-header"
        >
          Add Palette To Project: 
        </p>
        <div
          class="project-selector-container"
        >
          <label
            class="palette-form-label"
            for="select-project"
          >
            Select Project: 
          </label>
          <select
            class="project-selector"
            data-testid="select-project"
            id="select-project"
          >
            <option
              value=""
            >
              Please Select
            </option>
            <option
              value="test hahahah"
            >
              test hahahah
            </option>
            <option
              value="test"
            >
              test
            </option>
          </select>
        </div>
        <input
          class="palette-name-input"
          data-testid="palette-name-input"
          maxlength="30"
          placeholder="Palette Name"
          type="text"
          value=""
        />
        <div
          class="project-form-error-container"
        >
          <p
            class="project-form-error"
            data-testid="project-form-error"
            hidden=""
          >
            Error: Please Enter a Name and Project
          </p>
        </div>
        <button
          class="select-palette-btn"
          type="button"
        >
          Save Palette
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="palette-form"
    >
      <p
        class="palette-form-header"
      >
        Add Palette To Project: 
      </p>
      <div
        class="project-selector-container"
      >
        <label
          class="palette-form-label"
          for="select-project"
        >
          Select Project: 
        </label>
        <select
          class="project-selector"
          data-testid="select-project"
          id="select-project"
        >
          <option
            value=""
          >
            Please Select
          </option>
          <option
            value="test hahahah"
          >
            test hahahah
          </option>
          <option
            value="test"
          >
            test
          </option>
        </select>
      </div>
      <input
        class="palette-name-input"
        data-testid="palette-name-input"
        maxlength="30"
        placeholder="Palette Name"
        type="text"
        value=""
      />
      <div
        class="project-form-error-container"
      >
        <p
          class="project-form-error"
          data-testid="project-form-error"
          hidden=""
        >
          Error: Please Enter a Name and Project
        </p>
      </div>
      <button
        class="select-palette-btn"
        type="button"
      >
        Save Palette
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
